drift-transport-netty/src/main/java/com/facebook/drift/transport/netty/client/DriftNettyMethodInvokerFactory.java

```diff
@@ -88,38 +97,54 @@ public DriftNettyMethodInvokerFactory(
         this.sslContextFactory = createSslContextFactory(true, factoryConfig.getSslContextRefreshTime(), group);
         this.defaultSocksProxy = Optional.ofNullable(factoryConfig.getSocksProxy());
 
-        ConnectionManager connectionManager = new ConnectionFactory(group, sslContextFactory, allocator, factoryConfig);
-        if (factoryConfig.isConnectionPoolEnabled()) {
-            connectionManager = new ConnectionPool(
-                    connectionManager,
-                    group,
-                    factoryConfig.getConnectionPoolMaxSize(),
-                    factoryConfig.getConnectionPoolMaxConnectionsPerDestination(),
-                    factoryConfig.getConnectionPoolIdleTimeout());
-        }
-        this.connectionManager = connectionManager;
+        connectionPoolMaintenanceExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("drift-connection-maintenance"));
+        connectionFactory = new ConnectionFactory(group, sslContextFactory, allocator, factoryConfig);
     }
 
     @Override
     public MethodInvoker createMethodInvoker(I clientIdentity)
     {
-        ConnectionParameters clientConfig = toConnectionConfig(clientConfigurationProvider.apply(clientIdentity));
+        DriftNettyClientConfig driftNettyClientConfig = clientConfigurationProvider.apply(clientIdentity);
+        ConnectionParameters clientConfig = toConnectionConfig(driftNettyClientConfig);
 
         // validate ssl context configuration is valid
         clientConfig.getSslContextParameters()
                 .ifPresent(sslContextParameters -> sslContextFactory.get(sslContextParameters).get());
 
+        ConnectionManager connectionManager = getConnectionManager(clientIdentity, driftNettyClientConfig);
         return new DriftNettyMethodInvoker(clientConfig, connectionManager, group);
     }
 
+    public ConnectionManager getConnectionManager(I clientIdentity, DriftNettyClientConfig driftNettyClientConfig)
+    {
+        boolean connectionPoolEnabled = firstNonNull(driftNettyClientConfig.getConnectionPoolEnabled(), factoryConfig.isConnectionPoolEnabled());
+        if (!connectionPoolEnabled) {
+            return connectionFactory;
+        }
+
+        int connectionPoolMaxSize = firstNonNull(driftNettyClientConfig.getConnectionPoolMaxSize(), factoryConfig.getConnectionPoolMaxSize());
+        int maxConnectionsPerDestination = firstNonNull(driftNettyClientConfig.getConnectionPoolMaxConnectionsPerDestination(), factoryConfig.getConnectionPoolMaxConnectionsPerDestination());
+        Duration connectionPoolIdleTimeout = firstNonNull(driftNettyClientConfig.getConnectionPoolIdleTimeout(), factoryConfig.getConnectionPoolIdleTimeout());
+
+        return connectionPools.computeIfAbsent(clientIdentity, key -> new ConnectionPool(
```

Comment:
are we always guaranteed to have ```clientIdentity```? Should we assert it before calling this?

Reply:
In production codebase it looks like we are guaranteed, but in tests it is not. I changed the map key to be optional